### PR TITLE
chore!: Remove partial backend feature

### PR DESCRIPTION
**Pull request by kevaundray**
# Description

Related to #3772 

## Problem\*

Resolves 

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: tooling/backend_interface/src/proof_system.rs [hunk 1]
```diff
@@ -39,12 +39,12 @@ impl Backend {
     /// If we cannot get a valid backend, returns the default backend which supports all the opcodes
     /// and uses Plonk with width 3
```
Comment (TomAFrench) [inline tooling/backend_interface/src/proof_system.rs:40]:
```suggestion
    /// If we cannot get a valid backend, returns Plonk with width 3
```

File: tooling/backend_interface/src/proof_system.rs [hunk 2]
```diff
@@ -39,12 +39,12 @@ impl Backend {
...
     /// The function also prints a message saying we could not find a backend
-    pub fn get_backend_info_or_default(&self) -> (Language, BackendOpcodeSupport) {
-        if let Ok(backend_info) = self.get_backend_info() {
-            (backend_info.0, backend_info.1)
+    pub fn get_backend_info_or_default(&self) -> Language {
+        if let Ok(language) = self.get_backend_info() {
+            language
         } else {
             log::warn!("No valid backend found, defaulting to Plonk with width 3 and all opcodes supported");
```
Comment (TomAFrench) [inline tooling/backend_interface/src/proof_system.rs:46]:
```suggestion
            log::warn!("No valid backend found, defaulting to Plonk with width 3");
```

File: tooling/nargo/src/ops/optimize.rs [hunk 3]
```diff
@@ -20,11 +18,10 @@ pub fn optimize_program(
 pub fn optimize_contract(
     contract: CompiledContract,
     np_language: Language,
-    is_opcode_supported: &impl Fn(&Opcode) -> bool,
 ) -> Result<CompiledContract, NargoError> {
```
Comment (TomAFrench) [inline tooling/nargo/src/ops/optimize.rs:21]: We can remove the `Result` here now

File: tooling/nargo/src/ops/compile.rs [hunk 4]
```diff
@@ -85,7 +79,7 @@ pub fn compile_program(
         };
 
     // Apply backend specific optimizations.
-    let optimized_program = crate::ops::optimize_program(program, np_language, is_opcode_supported)
+    let optimized_program = crate::ops::optimize_program(program, np_language)
         .expect("Backend does not support an opcode that is in the IR");
```
Comment (TomAFrench) [inline tooling/nargo/src/ops/compile.rs:83]:
```suggestion
    let optimized_program = crate::ops::optimize_program(program, np_language);
```

File: tooling/nargo/src/ops/compile.rs [hunk 5]
```diff
@@ -106,9 +99,8 @@ fn compile_contract(
             }
         };
 
-    let optimized_contract =
-        crate::ops::optimize_contract(contract, np_language, &is_opcode_supported)
-            .expect("Backend does not support an opcode that is in the IR");
+    let optimized_contract = crate::ops::optimize_contract(contract, np_language)
+        .expect("Backend does not support an opcode that is in the IR");
```
Comment (TomAFrench) [inline tooling/nargo/src/ops/compile.rs:103]:
```suggestion
    let optimized_contract = crate::ops::optimize_contract(contract, np_language);
```

**Comment by kevaundray** [comment]
We can now also remove the `backend` field from PreprocessedProgram in place of a program_width parameter

**Review by TomAFrench — COMMENTED** [review]
LGTM

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**